### DISCUSSION TITLE
Handle Mediocre Beta Kata

### DISCUSSION THREAD

**Discussion by Kacarott** [codewars/codewars.com#2736] [Ideas]
As of the writing of this post, there are `2141` existing beta kata. However I would guess that the number of these kata which actually have any chance of ever being approved, is less than 200. One of the reasons for this is that the beta system currently only handles very bad kata (with retirement) or good kata (with eventual approval). Any kata which "settles" into a mediocre level of satisfaction is fated to just stay in beta indefinitely.

My suggestion is a way to deal with these mediocre kata which have settled at a low enough satisfaction rate that they have little hope of ever reaching `80%`. The measuring metric would be the number of 'Very satisfied' reviews in a row, required to bring the overall satisfaction up to `80%`. The higher this number is, the less chance the kata has of ever being approvable, so an appropriate cut off point can be used to automatically retire kata whose chances are too low.

Some examples:
[Doubly Linked List](https://www.codewars.com/kata/527b40435529dc624f000257) currently has `60%` satisfaction, out of 10 reviews. It would need exactly 10 positive reviews in a row to bring it to 80% satisfaction, making it barely approvable. But considering its current review history, this seems *incredibly* unlikely.

[Moving element in an array](https://www.codewars.com/kata/53d51336c2c772aa43000724) sits higher at a `70%` satisfaction rating, which might make it seem better, however it already has 40 reviews. It would require 20 positive reviews in a row to make it barely approvable.

[Decimal to binary converter](https://www.codewars.com/kata/589c414a139f7ef9bc00003b) with a very high `78%` satisfaction, has however already 269 reviews! Gaining 2% satisfaction to make it approvable, would require another 26 positive reviews in a row!

My suggestion for a final metric would be, any beta with `>5` reviews should be automatically retired if the number of positive reviews required in a row to reach `80%` is `>=10`. This would mean a kata with 5 reviews should have `>40%` satisfaction, a kata with 10 reviews should have `>60%` satisfaction, with 20 reviews `>70%` etc. asymptotically approaching `80%`.

**Comment by kazk** [comment]
So, retire beta kata with more than 5 votes if getting `n` straight up votes is not enough to make it approvable.

I tried some `n` with the following definition:
```ruby
# `true` if even getting `n` straight up votes is not enough
# `required` depends on average assessed rank
def hopeless?(n)
  n_votes > 5 && (n_up_votes + 0.5*n_ok_votes + n) / (n_votes + n) < required/100.0
end
```

Checked 2176 beta kata:

| n | retires | % retired |
| --: | ---: | ---: |
1 | 816 | 37.5%
2 | 757 | 34.8%
3 | 707 | 32.5%
4 | 659 | 30.3%
5 | 618 | 28.4%
8 | 501 | 23.0%
10 | 427 | 19.6%
12 | 347 | 15.9%
16 | 249 | 11.4%
20 | 157 | 7.2%
30 | 70 | 3.2%
50 | 27 | 1.2%
100 | 11 | 0.5%
200 | 4 | 0.2%
250 | 1 | 0.0%

**Comment by kazk** [comment]
50 straight up votes are not enough for these:

1. [National Bath Safety Month](https://www.codewars.com/kata/57d14a9d95497e9912000267) by [PG1](https://www.codewars.com/users/561183de470bdf5eb6000022) (75%)
1. [All Star Code Challenge #4](https://www.codewars.com/kata/586405a639c5abc2db0000e8) by [shaddyjr](https://www.codewars.com/users/56e8a4e427a00dca06001a90) (70%)
1. [Holiday I - Temperature in Bali](https://www.codewars.com/kata/57e8e9df2aee49c0280009ab) by [PG1](https://www.codewars.com/users/561183de470bdf5eb6000022) (73%)
1. [All Star Code Challenge #5](https://www.codewars.com/kata/5864152183f7e6e834000160) by [shaddyjr](https://www.codewars.com/users/56e8a4e427a00dca06001a90) (76%)
1. [Holy cats](https://www.codewars.com/kata/58ebfa6ef7cda1a3d4000048) by [aweleshetu](https://www.codewars.com/users/56e66928988e85980d0004d5) (76%)
1. [Grasshopper - if/else review](https://www.codewars.com/kata/57227b32785220c82b000b93) by [danleavitt0](https://www.codewars.com/users/559aab853cef38cb790000a9) (75%)
1. [SQL Basics: Biggest in Each Column](https://www.codewars.com/kata/5946e1947325ad198e0000a7) by [PG1](https://www.codewars.com/users/561183de470bdf5eb6000022) (75%)
1. [Greetings ](https://www.codewars.com/kata/5901c9d463bf406d7c000107) by [sam-karis](https://www.codewars.com/users/5818a0acacf02b19c700025a) (74%)
1. [No Loops 2 - You only need one - Shortest Code Version](https://www.codewars.com/kata/57cc91361ccdc727f90000cf) by [PG1](https://www.codewars.com/users/561183de470bdf5eb6000022) (74%)
1. [Stop worrying!](https://www.codewars.com/kata/58555e3db45c01bada0001e0) by [Samatar26](https://www.codewars.com/users/57fba83951a2213ea900004d) (72%)
1. [All Star Code Challenge #26](https://www.codewars.com/kata/5866d90edbca9ae420000145) by [shaddyjr](https://www.codewars.com/users/56e8a4e427a00dca06001a90) (70%)
1. [How bigga ma pizza ?](https://www.codewars.com/kata/5e9c06f95ea5b30033903194) by [dinglemouse](https://www.codewars.com/users/56d80ba911262d87be000c03) (74%)
1. [Climb Time - Can you send it? ](https://www.codewars.com/kata/59580a14838112d91f000174) by [hollyewhite](https://www.codewars.com/users/58db13a716a2a444e2000203) (65%)
1. [Is it Golden?](https://www.codewars.com/kata/585ade78f59b3c83c1000059) by [DiegoSalazar](https://www.codewars.com/users/54b948c1ac1bcdf42a000b5c) (72%)
1. [Counter (React)](https://www.codewars.com/kata/5f266f1f0dc7f30033d7b01d) by [Catherine1988](https://www.codewars.com/users/5ca21fa6a6730a001061280d) (53%)
1. [NOP](https://www.codewars.com/kata/5842ceb013ee5048350000bd) by [dinglemouse](https://www.codewars.com/users/56d80ba911262d87be000c03) (64%)
1. [Custom hook (React)](https://www.codewars.com/kata/5f2723a24e82120016ac0212) by [Catherine1988](https://www.codewars.com/users/5ca21fa6a6730a001061280d) (70%)
1. [Beam me up Statey using ReactJS](https://www.codewars.com/kata/5ac37e32b4218298130001aa) by [richkotze](https://www.codewars.com/users/58cd1793bd0fd320c1000075) (53%)
1. [Uncurrying](https://www.codewars.com/kata/56032c3deea6ebca2900003c) by [surtich](https://www.codewars.com/users/52ea969ecf476b78eb0004b7) (66%)
1. [Find twice of a number's value](https://www.codewars.com/kata/59357d303864834c1a000026) by [setkyar](https://www.codewars.com/users/538effaef4d60fa3f200063e) (59%)
1. [Find a number increase](https://www.codewars.com/kata/5934ef21b2defb45df000026) by [setkyar](https://www.codewars.com/users/538effaef4d60fa3f200063e) (57%)
1. [Basic Statement Coverage in Unit Testing](https://www.codewars.com/kata/5fad5be0ff1ef6000f928448) by [drchangliu](https://www.codewars.com/users/5f9f25496825190023b2aa75) (56%)
1. [Class extension model with mixin ability.](https://www.codewars.com/kata/52b2fcc1c65ea44b76000701) by [.blue](https://www.codewars.com/users/52a84ae54686d7889d0004a1) (55%)
1. [The Kingdom of Squares](https://www.codewars.com/kata/5ffc412fe72bac002b491214) by [Jari Jarvis](https://www.codewars.com/users/5ff98295c36c9900262cd329) (53%)
1. [[Code Golf] Sum of Minimums](https://www.codewars.com/kata/623f0be645722791e95bb23b) by [mubashir1](https://www.codewars.com/users/5f4d004618330d002e26481d) (54%)
1. [Random string](https://www.codewars.com/kata/55d92bf7934aff33c20000a2) by [sunboshan](https://www.codewars.com/users/55d7e77cdb4b5e7959000006) (47%)
1. [NHA JS1 Variables - Declaration & Assignment](https://www.codewars.com/kata/57e6dd0b6603f63e4f0000fd) by [lpaglione](https://www.codewars.com/users/57e680a52827a68b050001e2) (42%)

**Comment by kazk** [comment]
`n = 10` will retire all kata listed on https://github.com/codewars/content-issues/wiki/List-of-Mediocre-Beta-Kata

What do you think?

**Reply by Kacarott** [reply]
I think that `n = 10` would be a good threshold. With a quick scan over the list, I don't see any particular kata standing out as "underrated" or worth saving.

**Comment by hobovsky** [comment]
I like the general idea, but I am not sure it covers all necessary/useful areas:

- How would the retirement be triggered? Would there be some mechanism introduced, or is it one-off thing, or we ask admins to run a query periodically?
- Do we need to satisfy the hoarder's mindset many of us seem to display? I think one problem with removal of betas is a hope that _"it could be fixed one day"_, because the idea is nice and  execution is bad. Or that _I might  not like it, but others apparently do"_. After getting unpublished or retired, kata becomes virtually inaccessible. Is it a problem? It is not for me, but I think it could be for some.
- What about approved kata of low satisfaction? Do we want to handle these somehow? It's probably out of scope of this issue, but maybe it would be nice to have some way to handle also approved kata of low satisfaction, with many issues?

**Reply by Kacarott** [reply]
My intent with starting this discussion was for it to be an automatic process, essentially just an addition to the existing auto-retire thresholds. I think it is more valuable as an automatic process rather than a one-off thing, since in the future it will mean that it is always worthwhile solving and rating beta katas, since it will drive them closer to approval or retirement.

I don't think that we should be trying to satisfy hoarders mindset. The initial batch of retired kata could be kept (for a while) in the list for people's reference, and in the future if people are aware of the new retirement rule, then they can keep track of kata they want to save (via collections or bookmarks)

Your third point I think is outside the scope of this issue, as you said.

**Reply by kazk** [reply]
> How would the retirement be triggered?

Same as existing auto-retire.

> Do we need to satisfy the hoarder's mindset many of us seem to display? I think one problem with removal of betas is a hope that "it could be fixed one day", because the idea is nice and execution is bad. 

This is why I tried different `n`. `n = 50` is pretty much impossible to get it approved unless we reset the votes.
`n = 10` may include some that can be salvaged by fixing issues and removing some outdated votes.

> What about approved kata of low satisfaction?

We should start a separate discussion.

**Reply by hobovsky** [reply]
> This is why I tried different `n`. `n = 50` is pretty much impossible to get it approved unless we reset the votes.
> `n = 10` may include some that can be salvaged by fixing issues and removing some outdated votes.

I think this is the case with lolisa's HAskell  kata, which got recycled at some point. But still, even if somewhat strict, I like the threshold of 10. With current amount of betas, I prefer to cut off more rather than few. I am fine with 10.

**Comment by hobovsky** [comment]
Another potential problem: currently, it's difficult to establish how far from approval/retirement a kata is. Voter's voting power is a hidden stat which impacts satisfaction rating, but is not exposed anywhere. I wanted to propose (or create with a user script) some kind of an indicator which shows that "This kata can be approved after receiving X green votes, and retired after Y red votes" but it's not possible to get the data anywhere.
Would such indicator be useful?

**Reply by Kacarott** [reply]
Might be nice to have this, perhaps as a popup when hovering over the satisfaction rate?

**Comment by ejini6969** [comment]
I'm not sure if 10 is the right threshold, maybe abit too strict. I prefer 15, but that's just my opinion. I suggest we not do this auto-retire thing first and glance through the list 1-by-1, starting from the 60% threshold. Some of those kata are worth saving, like myjinxin's thinking & testing series, as well as bubbler's katas. But for those beyond 60%, they can be retired or sent back to draft imo due to being exact / mixture of multiple tasks directly or indirectly.

**Reply by Kacarott** [reply]
There is only one myjinxin kata in the list, and have you actually tried it? Or are you just assuming its worth saving based on the author. I am attempting it now and am finding it quite terrible, there is a good reason it has only 50% satisfaction.

The Bubbler kata also has low satisfaction, and seems to be essentially a duplicate of a different Bubbler kata which is already approved.

**Reply by Kacarott** [reply]
And besides, my suggestion would be for this list of kata to be left up, so that people can still look through and access the kata after retirement. Then if someone wants to save a kata they can access it and copy the good parts into a fresh kata.

**Reply by hobovsky** [reply]
I would propose to add another filter to kata search, to be able to search for retired kata. Maybe additional status, aside of "Approved" and "Beta"?
Potentially coupled with subcriterion of "Retirement reason" (retired by admin, or by reaching low satisfaction level, or by being "hopeless"), just like "Beta status" now, it would be possible for whoever enjoys this to search though retired, bad kata. Adding this to search mechanism has an additional advantage of being able to apply other filters and sorts too: filter by author, sort by satisfaction, rank (if assigned), etc. If someone finds a kata worth redeeming, they can fix it and ask privileged users to restore it.

I'd hope that making retired kata discoverable would possibly fix the problem of hoarding.

**Reply by Blind4Basics** [reply]
maybe restrict the search to menders/mods/admins?

**Reply by ejini6969** [reply]
I agree with approach proposed by hobs (considering the fact that most of the katas have issues raised and authors were alr long gone or unwilling to fix them !) And all this should be limited to content maintainers only as there will be abundant of copied solutions here and there if it's available to public! But imo we need more opinions from other PUs so this list should be kept as such until then, when retirement takes place!

**Reply by kazk** [reply]
I wouldn't want retired kata to be accessible too easily, but will allow menders/mods/admins to list them. I don't want average users to see them because retired kata can't be completed, and it might not even work at all because it hasn't been updated.

We currently have 2106 retired kata.

**Reply by hobovsky** [reply]
> I'm not sure if 10 is the right threshold, maybe abit too strict. I prefer 15, but that's just my opinion.

"10 upvotes away" might seem restrictive, but it doesnt necessarily have to be: note that this formula gives kata which need 10 green votes straight and nothing else, what, I believe, would effectively translate into a need for significantly many many more solutions and voters: if a kata does not meet the threshold already, it's highly probable that some users will vote yellow or red, effectively increasing the requirement.

**Comment by kazk** [comment]
Added the condition with `n = 10`, so it should auto-retire when the vote is updated by someone. I can later run a script to retire all beta matching the condition as well.

https://www.codewars.com/kata/retired can be used to search retired kata. This page is only for mods and menders.
